demos/portal-embed-demo/public/js/ttt/audio-controller.js

```diff
@@ -0,0 +1,369 @@
+        this._createNewAudio(audioId);
+        this.playlist = [audioId];
+        this.playlistIndex = 0;
+        this.isPlaying = false;
+
+        // Add event listeners
+        this._hookEvents();
+    }
+
+    /**
+     * CSS
+     */
+    get css() {
+        return `
+            :host {
+                width: 100%;
+                height: 80px;
+                background-color: slategray;
+                position: fixed;
+                bottom: 0px;
+                z-index: 999;
+                display: block;
+            }
+            .controllers {
+                margin: 0 auto;
+                height:80%;
+                width:180px;
+            }
+            .controller {
+                width: 60px;
+                height: 80%;
+                text-align: center;
+                float: left;
+            }
+            .controller img{
+                height: 50%;
+                cursor: pointer;
+            }
+            .flip-horizontal {
+                transform: scale(-1, 1);
+            }
+            .title {
+                width: 100%;
+                height: 20%;
+                text-align: center;
+                margin-top: 8px;
+                font-size: 0.8em;
+                color: #FFF;
+            }
+            .timeline {
+                width: 90%;
+                max-width: 550px;
+                border-top: 1px solid #FFF;
+                margin: 8px auto 3px auto;
+                cursor: pointer;
+            }
+            .axis{
+                width: 10px;
+                height: 10px;
+                background-color: #FFF;
+                position: relative;
+                left: 0px;
+                top: -5px;
+                border-radius: 50%;
+                cursor: pointer;
+            }
+            .prev {
+                opacity: 0.3;
+            }
+            .next {
+                opacity: 0.3;
+            }
+        `;
+    }
+
+    /**
+     * HTML template
+     */
+    get template() {
+        return `
+            <div class='title'></div>
+            <div class='timeline'>
+                <div class='axis'></div>
+            </div>
+            <div class='controllers'>
+                <div class='controller'>
+                    <img class='prev flip-horizontal' src='/img/next.png'>
+                </div>
+                <div class='controller'>
+                    <img class='play' src='/img/play.png'>
+                </div>
+                <div class='controller'>
+                    <img class='next' src='/img/next.png'>
+                </div>
+            </div>
+        `
+    }
+
+    /**
+     * Controlling play button
+     */
+    play() {
+        this.playButton.src = '/img/pause.png';
+        const startAxis = _ => {
+            // Start the progress bar
+            const left = this.timeline.offsetWidth;
+            this.axis.style.transition = `${this.durationSec}s linear`;
```

Comment:
I don't think this timing is correct if we've paused the audio half-way through the file. The progress bar will progress such that it would reach the end after `this.durationSec` seconds, but only `this.durationSec / 2` actually remain at that point. This should probably be set based on the actual seek position.

Reply:
It actually stops when pausing.
```javascript
pause() {
  ...
  // Stop the progress bar
  this.progressBar.style.left = `${this._getProgressBarCurrentLeft()}px`;
}
```
I'm wondering how I can implement using the actual seek position. I wasn't sure if we should use setInterval for this (wondering the performance cost of scheduling running a script per second). If there was some observables I can use, that could be another option but could you kindly advice me here?
```javascript
const updateProgressBar = _ => {
  this.progressBar.style.left = this.timeline.offsetWidth * this.audio.currentTime/this.audio.duration;
}
setInterval(updateProgressBar, 1000);
``

Reply:
> It actually stops when pausing.
> 
> ```js
> pause() {
>   ...
>   // Stop the progress bar
>   this.progressBar.style.left = `${this._getProgressBarCurrentLeft()}px`;
> }
> ```

Right, but what I mean is (even without accounting for audio buffering and other possible sources of delay): suppose the audio clip is 30 seconds long.1
1. Click play. 30-second transition starts (`this.durationSec`).
2. Wait 15 seconds.
3. Click pause. Transition pauses; style is kept at 50% of the progress bar width.
4. Click play. 30-second transition starts (`this.durationSec`).
5. Wait 15 seconds.
6. Audio clip finishes, but progress bar is only 75% to the right (50% of the transition from 50% to 100% has completed).

> I'm wondering how I can implement using the actual seek position. I wasn't sure if we should use setInterval for this (wondering the performance cost of scheduling running a script per second). If there was some observables I can use, that could be another option but could you kindly advice me here?
> 
> ```js
> const updateProgressBar = _ => {
>   this.progressBar.style.left = this.timeline.offsetWidth * this.audio.currentTime/this.audio.duration;
> }
> setInterval(updateProgressBar, 1000);
> ``
> ```

I believe this is what the [`timeupdate` event](https://developer.mozilla.org/en-US/docs/Web/API/HTMLMediaElement/timeupdate_event) is for (though I've never used it).

Reply:
THAT IS SO TRUE. I ended up implementing something like this: 
```javascript
        const startProgressBar = _ => {
            // Start the progress bar
            this.isPlaying = true;
            this.audio.addEventListener('timeupdate', evt => {
                this.progressBar.style.left = 
                    `${this.timeline.offsetWidth * this.audio.currentTime/this.audio.duration}px`;
            });
        }
```
And I get to delete `handleActivation` and ` _getProgressBarCurrentLeft` and remove all the unnecessary complexity!!!!!

Thank you!🙇🙇🙇

Reply:
Nice!

We can probably attach the `timeupdate` listener immediately and just leave it on. If we add the listener every time we start playing (which is what it seems like from the snippet), we'll eventually end up with a ton of copies of the listener (which won't break anything functionally but will get progressively slower over time).